src/main/java/no/unit/nva/doi/CrossRefClient.java

```diff
@@ -0,0 +1,122 @@
+import java.net.http.HttpClient;
+import java.net.http.HttpRequest;
+import java.net.http.HttpResponse;
+import java.net.http.HttpResponse.BodyHandlers;
+import java.time.Duration;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.ExecutionException;
+import javax.ws.rs.BadRequestException;
+import javax.ws.rs.NotFoundException;
+import javax.ws.rs.core.MediaType;
+import org.apache.http.HttpHeaders;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.utils.URIBuilder;
+import org.apache.http.client.utils.URLEncodedUtils;
+
+public class CrossRefClient {
+
+    public static final String CROSSREF_LINK = "https://api.crossref.org";
+    public static final String WORKS = "works";
+
+    public static final int TIMEOUT_DURATION = 30;
+    public static final String COULD_NOT_FIND_ENTRY_WITH_DOI = "Could not find entry with DOI:";
+    private final transient HttpClient httpClient;
+    private transient LambdaLogger logger;
+
+    public CrossRefClient() {
+        this(HttpClient.newHttpClient());
+    }
+
+    public CrossRefClient(HttpClient httpClient) {
+        this.httpClient = httpClient;
+    }
+
+    public Optional<String> fetchDataForDoi(String doiIdentifier) throws URISyntaxException {
+        URI targetUri = createTargetUrl(doiIdentifier);
+        return fetchJson(targetUri);
+    }
+
+    private Optional<String> fetchJson(URI doiUri) {
+        HttpRequest request = createRequest(doiUri);
+        try {
+            return Optional.ofNullable(getFromWeb(request));
+        } catch (InterruptedException |
+            ExecutionException |
+            NotFoundException |
+            BadRequestException e) {
+            System.out.print(e.getMessage());
+            logger.log(e.getMessage());
+            return Optional.empty();
+        }
+
+    }
+
+    private HttpRequest createRequest(URI doiUri) {
+        return HttpRequest.newBuilder(doiUri)
+                          .header(HttpHeaders.CONTENT_TYPE, MediaType.APPLICATION_JSON)
+                          .timeout(Duration.ofSeconds(TIMEOUT_DURATION))
+                          .GET()
+                          .build();
+    }
+
+    private String getFromWeb(HttpRequest request)
+        throws InterruptedException, ExecutionException {
+        HttpResponse<String> response = httpClient.sendAsync(request, BodyHandlers.ofString())
+                                                  .get();
+        if (responseIsSuccessful(response)) {
+            return response.body();
+        } else {
+            return handleError(request, response);
+        }
+    }
+
+    private String handleError(HttpRequest request, HttpResponse<String> response) {
+        if (response.statusCode() == HttpStatus.SC_NOT_FOUND) {
+            throw new NotFoundException(COULD_NOT_FIND_ENTRY_WITH_DOI + request.uri().toString());
+        }
+        throw new BadRequestException("Something went wrong");
```

Comment:
Should we log the other statusCode or something from the respons body?

Reply:
Fixed